crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -832,6 +840,228 @@ impl From<ProviderError> for EthFilterError {
+            }
+
+            let (_idx, header) = &self.headers[self.current_idx];
+            let header_idx = self.current_idx; // save index before incrementing
+            self.current_idx += 1;
+
+            // calculate block hash
+            let block_hash = if header_idx + 1 < self.headers.len() {
+                // use parent hash of next header to get current header's hash
+                self.headers[header_idx + 1].1.parent_hash()
+            } else {
+                // last header, need to fetch hash from provider
+                self.filter_inner
+                    .provider()
+                    .block_hash(header.number())?
+                    .ok_or_else(|| ProviderError::HeaderNotFound(header.number().into()))?
+            };
+
+            // try cache first for performance
+            if let Some((receipts, maybe_block)) =
+                self.filter_inner.eth_cache().get_receipts_and_maybe_block(block_hash).await?
+            {
+                return Ok(Some((receipts, maybe_block, header.clone(), block_hash)));
+            }
+
+            // cache miss - fallback to storage to ensure correctness
+            if let Some(receipts) =
+                self.filter_inner.provider().receipts_by_block(block_hash.into())?
+            {
+                let maybe_block = self
+                    .filter_inner
+                    .provider()
+                    .block(block_hash.into())?
+                    .and_then(|block| block.try_into_recovered().ok())
+                    .map(Arc::new);
+
+                return Ok(Some((Arc::new(receipts), maybe_block, header.clone(), block_hash)));
+            }
+        }
+    }
+}
+
+impl<
+        Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+    > RangeBlockMode<Eth>
+{
+    async fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            Arc<Vec<ProviderReceipt<Eth::Provider>>>,
+            Option<Arc<reth_primitives_traits::RecoveredBlock<ProviderBlock<Eth::Provider>>>>,
+            <Eth::Provider as HeaderProvider>::Header,
+            B256,
+        )>,
+        EthFilterError,
+    > {
+        loop {
+            // if we have current headers to process
+            if let Some(ref headers) = self.current_headers {
+                while self.current_header_idx < headers.len() {
+                    let header = &headers[self.current_header_idx];
+                    self.current_header_idx += 1;
+
+                    // calculate block hash
+                    let block_hash = if self.current_header_idx < headers.len() {
+                        headers[self.current_header_idx].parent_hash()
+                    } else {
+                        self.filter_inner
+                            .provider()
+                            .block_hash(header.number())?
+                            .ok_or_else(|| ProviderError::HeaderNotFound(header.number().into()))?
+                    };
```

Comment:
we have this twice, in both modes, I think we can move this to the callsite and then move `SealedHeader<>` into the modes?